lightwood/helpers/torch.py

```diff
@@ -19,3 +22,49 @@ def concat_vectors_and_pad(vec_list, max_):
 def average_vectors(vec_list):
     assert len(vec_list) > 0
     return torch.cat([emb[None] for emb in vec_list], dim=0).mean(0)
+
+
+class LightwoodAutocast:
+    """
+    Equivalent to torch.cuda.amp.autocast, but checks device compute capability
+    to activate the feature only when the GPU has tensor cores to leverage AMP.
+    """
+    def __init__(self, enabled=True):
+        self.major = 0  # GPU major version
+        torch_version = [int(i) for i in torch.__version__.split('.')[:-1]]
+
+        if not enabled or not torch.cuda.is_available() or torch_version[0] < 1 or torch_version[1] < 6:
+            self._enabled = False
+        else:
+            device, _ = get_devices()
+            if device.type == 'cuda':
+                self.major, _ = torch.cuda.get_device_capability(device)
+                self._enabled = enabled
+            else:
+                self._enabled = False  # gpu is available but cpu is forced
+
+        self.prev = self._enabled
+
+    def __enter__(self):
+        # tensor cores only exist from 7 onwards
+        # if this is not the case, then AMP is unnecessary overhead
+        if self.major > 6:
+            self.prev = torch.is_autocast_enabled()
```

Comment:
Why does this need to be checked with every `__enter__` ?
Why not check it once in the `__init__` and then just use `self._enabled` here, and never even use the `self.prev` variable.

Reply:
Good point, will fix